### PR TITLE
Authz

### PR DESCRIPTION
**Pull request by vuong177**
Closes: #2099 

## What is the purpose of the change
 -  Apply API changes from https://github.com/osmosis-labs/cosmos-sdk/pull/224 
## Brief Changelog

*(for example:)*
  -  `RegisterLegacyAminoCodec` in codec.go in every module. This allows the `x/authz` module to properly serialize and de-serialize `MsgExec` instances using Amino
## Testing and Verifying
 - Add golang test in msgs_test.go. This test confirms authz serialize and de-serializes working right.
 - I created a testnet with the current mainnet state. My ledger can sign authz msg in this testnet.

### CODE REVIEW

File: x/gamm/types/msgs_test.go [hunk 1]
```diff
@@ -867,3 +872,148 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
 		}
 	}
 }
+
+// Test authz
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:876]: Please elaborate elaborate here that this test is meant more so to test the authz than the gamm message

File: x/gamm/types/msgs_test.go [hunk 2]
```diff
@@ -867,3 +872,148 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
...
+	someDate := time.Date(1, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	testCases := []struct {
+		expectedGrantSignByteMsg   string
+		expectedRevokeSignByteMsg  string
+		expectedExecStrSignByteMsg string
+		gammMsg                    sdk.Msg
+	}{
+		{
+			expectedGrantSignByteMsg:   `{"account_number":"1","chain_id":"foo","fee":{"amount":[],"gas":"0"},"memo":"memo","msgs":[{"type":"cosmos-sdk/MsgGrant","value":{"grant":{"authorization":{"type":"cosmos-sdk/GenericAuthorization","value":{"msg":"/osmosis.gamm.v1beta1.MsgExitSwapShareAmountIn"}},"expiration":"0001-01-01T02:01:01.000000001Z"},"grantee":"cosmos1def","granter":"cosmos1abc"}}],"sequence":"1","timeout_height":"1"}`,
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:891]: Is there a way to have this represented in a struct format and the unmarshal to json string if needed?

It's hard to follow this in a string format.

File: x/gamm/types/msgs_test.go [hunk 3]
```diff
@@ -867,3 +872,148 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
...
+
+// Test authz
+func TestAuthzMsg(t *testing.T) {
+	appParams.SetAddressPrefixes()
+	pk1 := ed25519.GenPrivKey().PubKey()
+	addr1 := sdk.AccAddress(pk1.Address()).String()
+	coin := sdk.NewCoin("stake", sdk.NewInt(1))
+	someDate := time.Date(1, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	testCases := []struct {
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:884]: can we have names for the test cases please?

File: x/gamm/types/msgs_test.go [hunk 4]
```diff
@@ -867,3 +872,148 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
...
+				TokenInMaxAmount: sdk.NewInt(1),
+			},
+		},
+	}
+	for _, tc := range testCases {
+		// Authz: Grant Msg
+		typeURL := sdk.MsgTypeURL(tc.gammMsg)
+		grant, err := authz.NewGrant(someDate, authz.NewGenericAuthorization(typeURL), someDate.Add(time.Hour))
+		require.NoError(t, err)
+		msgGrant := &authz.MsgGrant{Granter: "cosmos1abc", Grantee: "cosmos1def", Grant: grant}
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:998]: can we extract `cosmos1abc` and `cosmos1def` to constants instead of hardcoding in multiple places, please?

File: x/superfluid/types/msg_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	appParams "github.com/osmosis-labs/osmosis/v10/app/params"
+
+	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto/ed25519"
+)
+
+func TestAuthzMsg(t *testing.T) {
```
Comment (p0mvn) [inline x/superfluid/types/msg_test.go:19]: Please add a comment describing the test

File: x/superfluid/types/msg_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	someDate := time.Date(1, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	testCases := []struct {
+		expectedGrantSignByteMsg   string
+		expectedRevokeSignByteMsg  string
+		expectedExecStrSignByteMsg string
+		msg                        sdk.Msg
+	}{
+		{
+			expectedGrantSignByteMsg:   `{"account_number":"1","chain_id":"foo","fee":{"amount":[],"gas":"0"},"memo":"memo","msgs":[{"type":"cosmos-sdk/MsgGrant","value":{"grant":{"authorization":{"type":"cosmos-sdk/GenericAuthorization","value":{"msg":"/osmosis.superfluid.MsgLockAndSuperfluidDelegate"}},"expiration":"0001-01-01T02:01:01.000000001Z"},"grantee":"cosmos1def","granter":"cosmos1abc"}}],"sequence":"1","timeout_height":"1"}`,
```
Comment (p0mvn) [inline x/superfluid/types/msg_test.go:33]: Similar comment about making these structs and unmarshalling, if possible

File: x/tokenfactory/types/msgs_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	appParams "github.com/osmosis-labs/osmosis/v10/app/params"
+
+	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto/ed25519"
+)
+
+func TestAuthzMsg(t *testing.T) {
```
Comment (p0mvn) [inline x/tokenfactory/types/msgs_test.go:19]: ditto: comment

File: x/tokenfactory/types/msgs_test.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	someDate := time.Date(1, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	testCases := []struct {
+		expectedGrantSignByteMsg   string
+		expectedRevokeSignByteMsg  string
+		expectedExecStrSignByteMsg string
+		msg                        sdk.Msg
+	}{
+		{
+			expectedGrantSignByteMsg:   `{"account_number":"1","chain_id":"foo","fee":{"amount":[],"gas":"0"},"memo":"memo","msgs":[{"type":"cosmos-sdk/MsgGrant","value":{"grant":{"authorization":{"type":"cosmos-sdk/GenericAuthorization","value":{"msg":"/osmosis.tokenfactory.v1beta1.MsgCreateDenom"}},"expiration":"0001-01-01T02:01:01.000000001Z"},"grantee":"cosmos1def","granter":"cosmos1abc"}}],"sequence":"1","timeout_height":"1"}`,
```
Comment (p0mvn) [inline x/tokenfactory/types/msgs_test.go:33]: ditto: structs instead of strings

**Review by p0mvn — COMMENTED** [review]
Awesome work @vuong177 . Thank you so much for the tests, including on the testnet.

Had some minor comments. Please take a look

**Comment by p0mvn** [comment]
Hi @vuong177 . Have you gotten a chance to look at my comments? Would love to help you get this over the finish line as we are prioritizing this feature for the current sprint completion

File: x/gamm/types/msgs_test.go (hunk 2 above)
Comment (vuong177) [reply x/gamm/types/msgs_test.go:891]: > Is there a way to have this represented in a struct format and the unmarshal to json string if needed?
> 
> It's hard to follow this in a string format.

I think it's a bit complex. Why we don't use pretty json in expectedGrantSignByteMsg. Like that
```
expectedGrantSignByteMsg :  `{
   "account_number":"1",
   "chain_id":"foo",
   "fee":{
      "amount":[
         
      ],
      "gas":"0"
   },
   "memo":"memo",
   "msgs":[
      {
         "type":"cosmos-sdk/MsgGrant",
         "value":{
            "grant":{
               "authorization":{
                  "type":"cosmos-sdk/GenericAuthorization",
                  "value":{
                     "msg":"/osmosis.gamm.v1beta1.MsgExitSwapShareAmountIn"
                  }
               },
               "expiration":"0001-01-01T02:01:01.000000001Z"
            },
            "grantee":"cosmos1def",
            "granter":"cosmos1abc"
         }
      }
   ],
   "sequence":"1",
   "timeout_height":"1"
}`
```
instead of use one line string

Comment (p0mvn) [reply x/gamm/types/msgs_test.go:891]: Sounds good, thanks for suggesting the alternative

**Comment by vuong177** [comment]
@p0mvn sorry for my late respnse. I'm making this test cleaner.

**Comment by p0mvn** [comment]
> @p0mvn sorry for my late response. I'm making this test cleaner.

No problem and no rush, just making sure that we are making progress here :+1:

**Comment by ValarDragon** [comment]
Woah thanks for this PR!

I think somethings off about having this many test vectors for authz in every module. Whats the exact authz property were trying to test for every message?

What I was hoping would be the case is that we can have a `suite.RunStandardMessageTest[M: sdk.Msg](msg M)`, and then authz having completeness could be tested there. But we don't need to care about the exact sign bytes, we can trust the authz module to get that correct

**Comment by vuong177** [comment]
@ValarDragon This test is for checking serialize and de-serialize in MsgExec. The current osmosis version don't `GetSignBytes` right. Example in a gamm msg :

Current version:
```
{
  "account_number": "1",
  "chain_id": "foo",
  "fee": {
    "amount": [],
    "gas": "0"
  },
  "memo": "memo",
  "msgs": [
    {
      "type": "cosmos-sdk/MsgExec",
      "value": {
        "grantee": "cosmos1def",
        "msgs": [
          {
            "pool_id": "1",
            "sender": "osmo14kva2pk8xkpkal0v46yp4vsn27pylayq0zdlls",
            "share_in_amount": "100",
            "token_out_denom": "test",
            "token_out_min_amount": "100"
          }
        ]
      }
    }
  ],
  "sequence": "1",
  "timeout_height": "1"
}
```
This PR :
```
{
  "account_number": "1",
  "chain_id": "foo",
  "fee": {
    "amount": [],
    "gas": "0"
  },
  "memo": "memo",
  "msgs": [
    {
      "type": "cosmos-sdk/MsgExec",
      "value": {
        "grantee": "cosmos1def",
        "msgs": [
          {
            "type": "osmosis/gamm/exit-swap-share-amount-in",
            "value": {
              "pool_id": "1",
              "sender": "osmo14kva2pk8xkpkal0v46yp4vsn27pylayq0zdlls",
              "share_in_amount": "100",
              "token_out_denom": "test",
              "token_out_min_amount": "100"
            }
          }
        ]
      }
    }
  ],
  "sequence": "1",
  "timeout_height": "1"
}
```

**Comment by ValarDragon** [comment]
What confuses is me is that there should be some generic validation logic for checking this -- if there wasn't, then there wouldn't have been any failures.

Can we just re-use that method (or write it), and then ensure `prior_deserialization(correct_authz_serialization(msg))` doesn't return an error?

**Comment by vuong177** [comment]
> What confuses is me is that there should be some generic validation logic for checking this -- if there wasn't, then there wouldn't have been any failures.
> 
> Can we just re-use that method (or write it), and then ensure `prior_deserialization(correct_authz_serialization(msg))` doesn't return an error?

Oh great idea. I think we can do that. It makes this test more sense! Thank you @ValarDragon !

**Comment by ValarDragon** [comment]
Awesome, thank you so much! That will become much more maintainable as well =)

File: x/gamm/types/msgs_test.go [hunk 9]
```diff
@@ -867,3 +871,132 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
...
+			require.NoError(t, err)
+
+			// Authz: Exec Msg
+			msgAny, _ := cdctypes.NewAnyWithValue(tc.gammMsg)
+			msgExec := authz.MsgExec{Grantee: mockGrantee, Msgs: []*cdctypes.Any{msgAny}}
+			execMsgByte := json.RawMessage(sdk.MustSortJSON(authzcodec.ModuleCdc.MustMarshalJSON(&msgExec)))
+			err = authzcodec.ModuleCdc.UnmarshalJSON(execMsgByte, &mockMsgExec)
+			require.NoError(t, err)
+			require.Equal(t, msgExec.Msgs[0].Value, mockMsgExec.Msgs[0].Value)
+		})
```
Comment (ValarDragon) [inline x/gamm/types/msgs_test.go:1000]: Can we move this all to `apptesting` and call it `TestMessageSerialization` ?

Comment (vuong177) [reply x/gamm/types/msgs_test.go:1000]: Sure !

Comment (vuong177) [reply x/gamm/types/msgs_test.go:1000]: Should I rename package in `msg_test.go` from `types` to `types_test` ? This is to avoid import cycle error.

Comment (ValarDragon) [reply x/gamm/types/msgs_test.go]: ah yeah, this makes sense to do

File: x/gamm/types/msgs_test.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/types/msgs_test.go:880]: I actually don't think names mattered here, its pretty clear & automatically derivable from the message type, but not really an issue

**Review by ValarDragon — APPROVED** [review]
Ty for doing this! Looks good to me!

Note to others, diff is only large due to the _testing package switch, which makes sense

Lets add a changelog and this is good to go?

**Comment by vuong177** [comment]
I added changelog. Thanks for your reviews !

File: app/apptesting/test_suite.go [hunk 10]
```diff
@@ -323,6 +329,44 @@ func CreateRandomAccounts(numAccts int) []sdk.AccAddress {
...
+	require.NoError(t, err)
+
+	// Authz: Revoke Msg
+	msgRevoke := authz.MsgRevoke{Granter: mockGranter, Grantee: mockGrantee, MsgTypeUrl: typeURL}
+	msgRevokeByte := json.RawMessage(sdk.MustSortJSON(authzcodec.ModuleCdc.MustMarshalJSON(&msgRevoke)))
+	err = authzcodec.ModuleCdc.UnmarshalJSON(msgRevokeByte, &mockMsgRevoke)
+	require.NoError(t, err)
+
+	// Authz: Exec Msg
+	msgAny, _ := cdctypes.NewAnyWithValue(msg)
```
Comment (p0mvn) [inline app/apptesting/test_suite.go:362]: Let's handle the error please

File: x/gamm/types/msgs_test.go [hunk 11]
```diff
@@ -34,97 +36,97 @@ func TestMsgSwapExactAmountIn(t *testing.T) {
...
+			msg: createMsg(func(msg gammtypes.MsgSwapExactAmountIn) gammtypes.MsgSwapExactAmountIn {
 				msg.TokenOutMinAmount = sdk.NewInt(0)
 				return msg
 			}),
 			expectPass: false,
 		},
 		{
 			name: "negative amount criteria",
-			msg: createMsg(func(msg MsgSwapExactAmountIn) MsgSwapExactAmountIn {
+			msg: createMsg(func(msg gammtypes.MsgSwapExactAmountIn) gammtypes.MsgSwapExactAmountIn {
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:129]: q: Would it make sense to split up pass-through changes into a separate PR so that we can backport it? 

Otherwise, backporting other work can get bottlenecked on this later

cc: @ValarDragon

File: x/gamm/types/msgs_test.go (hunk 3 above)
Comment (p0mvn) [reply x/gamm/types/msgs_test.go:880]: On another lookg, agreed

File: x/gamm/types/msgs_test.go [hunk 12]
```diff
@@ -867,3 +869,98 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
 		}
 	}
 }
+
+// Test authz serialize and de-serializes for gamm msg.
+func TestAuthzMsg(t *testing.T) {
+	appParams.SetAddressPrefixes()
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:875]: I wasn't sure why we need this line so I tried removing it and the test still passes
```suggestion
```

File: x/gamm/types/msgs_test.go [hunk 13]
```diff
@@ -867,3 +869,98 @@ func TestMsgExitSwapShareAmountIn(t *testing.T) {
...
+	pk1 := ed25519.GenPrivKey().PubKey()
+	addr1 := sdk.AccAddress(pk1.Address()).String()
+	coin := sdk.NewCoin("stake", sdk.NewInt(1))
```
Comment (p0mvn) [inline x/gamm/types/msgs_test.go:878]:
```suggestion
	coin := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(1))
```

**Review by p0mvn — COMMENTED** [review]
LGTM overall, nice work @vuong177 

I'm wondering though if it makes sense to separate the types package name change so that we can backport it to `v11.x`. Please let me know what you think

File: x/gamm/types/msgs_test.go (hunk 11 above)
Comment (vuong177) [reply x/gamm/types/msgs_test.go:129]: I agree with creating a new PR to change package name.

Comment (vuong177) [reply x/gamm/types/msgs_test.go:129]: I'm going to create it.

Comment (ValarDragon) [reply x/gamm/types/msgs_test.go:129]: ah thats a great point, should also be a really quick review + merge

Comment (p0mvn) [reply x/gamm/types/msgs_test.go:129]: Thank you @vuong177 , merged https://github.com/osmosis-labs/osmosis/pull/2316

**Comment by ValarDragon** [comment]
oh rip merge conflicts, can you try git merge main?

(Thanks for making the other PR off the osmosis repo, once this is on the repo, we can get these last steps handled with async round trip lol)

**Comment by vuong177** [comment]
yes. No conflict anymore!

**Comment by ValarDragon** [comment]
Thanks for making this!-
